.devcontainer/devcontainer.json

```diff
@@ -0,0 +1,57 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.223.0/containers/python-3
+{
+    "name": "Python 3",
+    "build": {
+        "dockerfile": "Dockerfile",
+        "context": "..",
+        "args": {
+            // Update 'VARIANT' to pick a Python version: 3, 3.10, 3.9, 3.8, 3.7, 3.6
+            // Append -bullseye or -buster to pin to an OS version.
+            // Use -bullseye variants on local on arm64/Apple Silicon.
+            "VARIANT": "3.10-bullseye",
+            // Options
+            "NODE_VERSION": "lts/*"
+        }
+    },
+    // Set *default* container specific settings.json values on container create.
+    "settings": {
+        "python.defaultInterpreterPath": "/usr/local/bin/python",
+        "python.linting.enabled": true,
+        "python.linting.pylintEnabled": true,
+        "python.formatting.autopep8Path": "/usr/local/py-utils/bin/autopep8",
+        "python.formatting.blackPath": "/usr/local/py-utils/bin/black",
+        "python.formatting.yapfPath": "/usr/local/py-utils/bin/yapf",
+        "python.linting.banditPath": "/usr/local/py-utils/bin/bandit",
+        "python.linting.flake8Path": "/usr/local/py-utils/bin/flake8",
+        "python.linting.mypyPath": "/usr/local/py-utils/bin/mypy",
+        "python.linting.pycodestylePath": "/usr/local/py-utils/bin/pycodestyle",
+        "python.linting.pydocstylePath": "/usr/local/py-utils/bin/pydocstyle",
+        "python.linting.pylintPath": "/usr/local/py-utils/bin/pylint",
+        "python.testing.pytestEnabled": true,
+        "restructuredtext.confPath": "${workspaceFolder}/docs",
+        "restructuredtext.pythonRecommendation.disabled": true
+    },
+    // Add the IDs of extensions you want installed when the container is created.
+    "extensions": [
+        "ms-python.python",
+        "ms-python.vscode-pylance",
+        "eamodio.gitlens",
+        "EditorConfig.EditorConfig",
+        "GitHub.vscode-pull-request-github",
+        "joaompinto.vscode-graphviz",
+        "mutantdino.resource-monitor",
+        "lextudio.restructuredtext",
+        "trond-snekvik.simple-rst"
+    ],
+    // Use 'forwardPorts' to make a list of ports inside the container available locally.
+    // "forwardPorts": [],
```

Comment:
VS Code is sort of magic. The `forwardPorts` config is to set up any ports you want enabled by default. The reason your thing worked is because VS Code will notice when a process starts listening on a port and will automatically forward that port for you in addition to the defaults. You may have noticed that when you clicked on `http://0.0.0.0:8000` your browser actually opened `http://localhost:8000`. It's not a security hole: the port is only forwarded to be exposed on your local machine, it's not exposing the port to anyone else. Pure convenience.

Reply:
If you want to disable that, by the way, the setting is `remote.autoForwardPorts`

Reply:
Thanks for explaining! Updated the comments about this in 9b770be.